Switch parsing and `find` to by default use only names in the current namespace

pchiusano opened the pull request:
This fixes #2988 

It does so by tweaking a couple places in `HandleInput` and `Backend` to use only names under the current namespace rather than searching globally. I added `find.global` as an extra option, and if `find` returns no results, there's now a hint to suggest using `find.global`.

<img width="665" alt="image" src="https://user-images.githubusercontent.com/11074/160302824-c06193e0-e923-4716-a001-3e04cc4e7e03.png">

## Interesting decisions

This doesn't try to do anything fancy like only load the current namespace into RAM. That would be cool, but could be done later and isn't a behavior change, just a memory optimization.

I also left the "pretty-printing names" alone - the pretty-printer is still given all names globally, and will pick globally unique suffixes. This might be overkill, but I was thinking this "degrades gracefully" if people have some namespaces that aren't self-contained. If I didn't do this, then the user would be shown hashes when viewing definitions that aren't within a self-contained namespace. 

## What now? 

I haven't played with this much, but it seems nice so far. I'm going to dogfood this branch for my daily Unison development and try to sand any rough edges. I'd encourage other folks to try the same if they want.

## Loose ends

We should probably change where `base` gets fetched - I'd put it in `.scratch.base` rather than `.base`.

And then we should update docs if/when this gets merged.

pchiusano committed:
Switch to self-contained namespace model

pchiusano committed:
Add find.global command. Tweak output messages to suggest it on no results.

aryairani commented on unison-cli/src/Unison/CommandLine/InputPatterns.hs:
```diff
@@ -399,9 +399,15 @@ viewByPrefix =
+findGlobal :: InputPattern
+findGlobal = find' "find.global" True
+
+find' :: String -> Bool -> InputPattern
```
Maybe a new type here instead of Bool?

aryairani replied:
IsVerbose, and Global vs Local?
What was isVerbose again anyway? `ls` vs `find`?

pchiusano replied:
`isVerbose` is whether to include hashes in the results.

There was an unused `Bool` flag that I just repurposed for `find` vs `find.global`. I'm not actually sure what the intent of it was... since it was totally ignored in `HandleInput`. :)

I can do some `newtype`s though.

pchiusano replied:
Tried out a new type, did not love it, went with type alias.

sullyj3 commented:
What about having find take an optional namespace argument? Then `find.global foo` becomes `find . foo`, and you simultaneously eliminate a lot of unnecessary cding around

pchiusano commented:
https://github.com/unisonweb/unison/issues/2988#issuecomment-1080650808

Might be a good tweak to make this work nicer. I like the idea of not needing to do anything fancy for the library install process and retaining the full DAG of library dependencies (which can then be used for name resolution) rather than flattening it.

pchiusano committed:
dependents command now correctly shows local dependents

didn't bother with dependents.global

pchiusano committed:
Sort dependents output by name, not by hash! And put name in first column, rather than hash.

pchiusano committed:
Merge remote-tracking branch 'origin/trunk' into topic/contained-ns

# Conflicts:
#	unison-cli/src/Unison/CommandLine/InputPatterns.hs

pchiusano committed:
Merge remote-tracking branch 'origin/trunk' into topic/contained-ns

pchiusano committed:
Merge remote-tracking branch 'origin/trunk' into topic/contained-ns

pchiusano commented:
I’ve been dogfooding this all week and it’s been great. No complaints. I don’t even notice it really, which is more than I could say for the old behavior.

I’m going to fix up the failing transcripts (reliant on old behavior) and merge. 

I might add the ranking of names idea based on ‘lib’ segments also, I think that will help make this even more scalable.

pchiusano committed:
Merge remote-tracking branch 'origin/trunk' into topic/contained-ns

pchiusano committed:
fix up transcripts

pchiusano committed:
Merge remote-tracking branch 'origin/trunk' into topic/contained-ns

pchiusano committed:
transcript refresh

ceedubs commented:
@pchiusano I've been using this branch for about an hour or two of Unison development, and so far it has felt right and I think that it will solve a lot of the common issues that I run into.

One observation: I think that we either want `names` to not be limited to the current namespace or add a `names.global` command to mirror `find.global`.

pchiusano committed:
names and names.global commands

pchiusano commented:
@ceedubs good call on `names` I've added `names.global` and a helpful hint suggesting its use.

pchiusano committed:
lib segment ranking

pchiusano committed:
tweak output messages + refresh transcripts

pchiusano committed:
refresh a transcript

pchiusano commented:
Okay, this is ready to go, will merge after CI. A couple final notes:

* I did add `names` and `names.global` commands, with `names` just giving aliases in the current namespace and `names.global` giving them globally.
* I implemented ranked suffix-based resolution based on number of `lib` segments https://github.com/unisonweb/unison/issues/2988#issuecomment-1080650808 via the function `Name.searchByRankedSuffix`.
* Added a bit of verbiage in `suffixes.md` transcript describing and exercising the new behavior.
* A number of transcripts had to be changed as they were reliant on being able to resolve names in some great aunt / uncle or cousin namespace. Nothing exciting here.

pchiusano commented on parser-typechecker/src/Unison/PrintError.hs:
```diff
@@ -500,7 +500,7 @@ renderTypeError e env src = case e of
             annotatedAsErrorSite src termSite,
             case expectedType of
-              Type.Var' (TypeVar.Existential {}) -> "\nThere are no constraints on its type."
+              Type.Var' (TypeVar.Existential {}) -> "\nThere are no constraints on its type. "
```
Missing a space in this message. It was super bugging me.

pchiusano commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -949,8 +949,8 @@ loop = do
                 fixupOutput = fmap Path.toName . HQ'.toHQ . Path.unsplitHQ
-            NamesI thing -> do
-              ns0 <- basicParseNames
+            NamesI global thing -> do
+              ns0 <- if global then pure basicPrettyPrintNames else basicParseNames
```
`NamesI` picks up a `IsGlobal` flag, which is used to control which scope names to use (the local parsing names, or the global pretty print names).

Maybe `basicPrettyPrintNames` should be renamed to `globalPrettyPrintNames` and `basicParseNames` to `localParseNames`?

pchiusano commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -963,7 +963,7 @@ loop = do
                     where
                       go r = (r, NamesWithHistory.typeName hqLength r printNames)
-              respond $ ListNames hqLength (toList types') (toList terms')
+              respond $ ListNames global hqLength (toList types') (toList terms')
```
`ListNames` picks up an `IsGlobal` flag as well, so it can print a helpful tip about using `names.global` or not, if the user has already just used `names.global`.

pchiusano commented:
> We should probably change where base gets fetched - I'd put it in .scratch.base rather than .base.

Actually I no longer agree with this. Where it is is fine. You can always just `fork .base .myproject.lib.base` if you want to make `base` available to your project.
